Kernel/Arch/riscv64/Interrupts.cpp

```diff
@@ -32,18 +36,78 @@ ErrorOr<u8> reserve_interrupt_handlers(u8)
     TODO_RISCV64();
 }
 
-void register_generic_interrupt_handler(u8, GenericInterruptHandler&)
+static void revert_to_unused_handler(u8 interrupt_number)
 {
-    TODO_RISCV64();
+    auto* handler = new UnhandledInterruptHandler(interrupt_number);
+    handler->register_interrupt_handler();
+}
+
+void register_generic_interrupt_handler(u8 interrupt_number, GenericInterruptHandler& handler)
+{
+    auto*& handler_slot = s_interrupt_handlers[interrupt_number];
+    if (handler_slot == nullptr) {
+        handler_slot = &handler;
+        return;
+    }
+    if (handler_slot->type() == HandlerType::UnhandledInterruptHandler) {
+        auto* unhandled_handler = static_cast<UnhandledInterruptHandler*>(handler_slot);
+        unhandled_handler->unregister_interrupt_handler();
+        delete unhandled_handler;
+        handler_slot = &handler;
+        return;
+    }
+    if (handler_slot->is_shared_handler()) {
+        VERIFY(handler_slot->type() == HandlerType::SharedIRQHandler);
+        static_cast<SharedIRQHandler*>(handler_slot)->register_handler(handler);
+        return;
+    }
+    if (!handler_slot->is_shared_handler()) {
+        if (handler_slot->type() == HandlerType::SpuriousInterruptHandler) {
+            // FIXME: Add support for spurious interrupts on riscv64
+            TODO_RISCV64();
+        }
+        VERIFY(handler_slot->type() == HandlerType::IRQHandler);
+        auto& previous_handler = *handler_slot;
+        handler_slot = nullptr;
+        SharedIRQHandler::initialize(interrupt_number);
+        VERIFY(handler_slot);
+        static_cast<SharedIRQHandler*>(handler_slot)->register_handler(previous_handler);
```

Comment:
Should probably be unregister_handler?

Reply:
This usage of register_handler has been there since bb73802b156baae9bf0264d5de169d761e0612a9. I think you are right in that it should be unregister_handler?

Reply:
In that case, I'd like someone else to take a look since I'm not actually familiar with the interrupt handling code.

Reply:
No, now that I look at this code again,
this is supposed to be a register_handler.
This code turns an IRQHandler into a SharedIRQHandler when register_generic_interrupt_handler gets called with an already used IRQ.
And the old and new irq handler get registered at the shared irq handler.

Reply:
I also added a note about stval always being 0.
We should also note that Sstvala required when we document the RISC-V port, next to other extension we reuquire/support.